### DISCUSSION TITLE
How to receive messages from the same queue by multiple customers?

### DISCUSSION THREAD

**Discussion by ZedZipDev** [rabbitmq/rabbitmq-dotnet-client#1121] [General]
I run the producer, it generates N messages, i see them on the dashboard. 
When I run a receiver it receive all messages from the queue and the queue is an empty.
```
        static void Receive(string QueName)
        {
            ConnectionFactory connectionFactory = new ConnectionFactory
            {
                HostName = HostName,
                UserName = UserName,
                Password = Password,
            };
            var connection = connectionFactory.CreateConnection();
            var channel = connection.CreateModel();
            channel.BasicQos(0, 1, false);
            MessageReceiver messageReceiver = new MessageReceiver(channel);
            channel.BasicConsume(QueName, false, messageReceiver);
            Console.WriteLine(" Press [enter] to exit.");
            Console.ReadLine();
        }

```
I need to have multiple producers which generate messages to the same queue.
And multiple customers receive messages from the queue. And messages will be deleted by queue TTL.
But now the 1st receiver gets all messages from the queue.
How can I do this?

**Comment by lukebakken** [comment]
Everything appears correct in your code. You should share a complete set of code I can compile and run for your producer and consumer application.

Are all consumers using the same `QueName` value?

Finally, please let us know the version of _all_ software you are using, including RabbitMQ and Erlang.

**Reply by ZedZipDev** [reply]
I use the Docker as a RabbitMQ server:
[PS D:\rabbitmq> docker run -d --hostname my-rabbit --name some-rabbit -p 15672:15672 -p 5672:5672 rabbitmq:3-management](url)

I have attached 2 projects: producer and receiver
[RabbitMQTest2Rcv.zip](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/7619410/RabbitMQTest2Rcv.zip)
[RabbitMQTest2Producer.zip](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/7619411/RabbitMQTest2Producer.zip)

**Reply by lukebakken** [reply]
Thank you. I will try to find time for this over the next week. I have added the code to the following repository and will be making changes there:

https://github.com/lukebakken/rabbitmq-dotnet-client-1121

**Reply by ZedZipDev** [reply]
Thank you, great

**Reply by ZedZipDev** [reply]
Yet another to this topic. 
I have made step by step what described here

https://www.tutlane.com/tutorial/rabbitmq/csharp-rabbitmq-fanout-exchange

and see: yes, the producer sends message to all fanout bound queues. Its ok. But when I run receiver (it reads from "Mumbai" queue as in article) it reads messages and queue "Mumbai" is empty. 
I.e. if another receiver tries to read from the same queue it cannot read any messages.
The 1st receiver reads all messages from queue and the queue is an empty.
But I'd like to any receiver can read messages from this queue and messages will be deleted by RabbitMQ server using TTL.
May be some Ack\NoAck to use in other manner?
Thank you.

**Reply by lukebakken** [reply]
So that we're working on the same code, I have added you as a collaborator to my repository -

https://github.com/lukebakken/rabbitmq-dotnet-client-1121

Could you please ensure the code there is what you are working on? I will be spending some time on it today or tomorrow.

**Comment by lukebakken** [comment]
Please see this commit:

https://github.com/lukebakken/rabbitmq-dotnet-client-1121/commit/ee9922f895d1b9dfe5635d72762500056f59d2df

I have simplified your code to work with a fanout exchange, and I added a README with instructions:

https://github.com/lukebakken/rabbitmq-dotnet-client-1121/blob/main/README.md

**Reply by ZedZipDev** [reply]
Thank you @lukebakken , I see: every customer has its own queue bound with fanout exchange. 
The producer generates messages via fanout exchange, they sent to all queues bound to amp.fanout.
Ok, it is little bit different but it is solution.
Each customer has its own queue bound to fanout exchange.

**Reply by ZedZipDev** [reply]
One question: here you create the queue:

```
 using (var connection = connectionFactory.CreateConnection())
            {
                using (var channel = connection.CreateModel())
                {
                    QueueDeclareOk queueDeclareOk = channel.QueueDeclare(queueName, true, false, false);
                    System.Diagnostics.Debug.Assert(queueDeclareOk.QueueName == queueName);
                    channel.QueueBind(queueName, "amq.fanout", string.Empty);

                    channel.BasicQos(0, 1, false);
                    var messageReceiver = new MessageReceiver(channel);
                    channel.BasicConsume(queueName, false, messageReceiver);

                    Console.WriteLine(" Press [enter] to exit.");
                    Console.ReadLine();
                }
            }
```

The second call of 
 QueueDeclareOk queueDeclareOk = channel.QueueDeclare(queueName, true, false, false);

returns data about exisitng queue, it is ok.
How to be if I want to create a queue with TTL?

If I use this parameter and do the 2nd call 
```
 var args = new Dictionary<string, object>();
 args.Add("x-message-ttl", 160000);
 QueueDeclareOk queueDeclareOk = channel.QueueDeclare(queueName, true, false, false,args);
```
it generates an error and closes the channel.

How to do this?

**Reply by lukebakken** [reply]
When you re-declare a queue that already exists, the parameters must match. My guess is that the queue already exists *without* TTL and you're trying to declare it *with* TTL.

The solution is to delete the queue first, or start with a fresh RabbitMQ instance.

**Reply by lukebakken** [reply]
https://www.rabbitmq.com/queues.html#property-equivalence
